#### Get started with TensorFlow

In [1]:
import tensorflow as tf
mnist = tf.keras.datasets.mnist

(x_train, y_train),(x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
  loss='sparse_categorical_crossentropy',
  metrics=['accuracy'])

model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test, y_test)


Epoch 1/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.3032 - accuracy: 0.9116
Epoch 2/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1451 - accuracy: 0.9573
Epoch 3/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1101 - accuracy: 0.9676
Epoch 4/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0876 - accuracy: 0.9734
Epoch 5/5
313/313 [==============================] - 0s 742us/step - loss: 0.0745 - accuracy: 0.9780


[0.07453450560569763, 0.9779999852180481]

### **Epoch 1/5**

**Output:**
```
1875/1875 [==============================] - 3s 1ms/step - loss: 0.3032 - accuracy: 0.9116
```

**Explanation:**
- **Epoch 1/5**: This indicates that the model is currently training on the first epoch out of a total of 5 epochs. An epoch refers to one complete pass through the entire training dataset.
- **1875/1875**: The training dataset is divided into batches, and each batch contains a subset of the training data. Here, 1875 batches are processed per epoch. The progress bar `[==============================]` visually indicates the progress through the batches.
- **3s**: The total time taken for this epoch to complete is 3 seconds.
- **1ms/step**: The average time taken to process each batch (step) is 1 millisecond.
- **loss: 0.3032**: The loss value after the first epoch is 0.3032. Loss is a measure of how well the model's predictions match the actual labels. A lower loss indicates better performance.
- **accuracy: 0.9116**: The accuracy of the model on the training data after the first epoch is 91.16%. Accuracy is the proportion of correctly predicted labels out of the total labels.

### **Epoch 2/5**

**Output:**
```
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1451 - accuracy: 0.9573
```

**Explanation:**
- The model is now on the second epoch.
- **2s**: The time taken for the second epoch is 2 seconds, which is slightly faster than the first epoch.
- **loss: 0.1451**: The loss has decreased to 0.1451, indicating that the model's predictions are improving.
- **accuracy: 0.9573**: The accuracy has increased to 95.73%, showing that the model is learning and performing better on the training data.

### **Epoch 3/5**

**Output:**
```
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1101 - accuracy: 0.9676
```

**Explanation:**
- The model is on the third epoch.
- **2s**: The time taken remains consistent at 2 seconds.
- **loss: 0.1101**: The loss has further decreased to 0.1101.
- **accuracy: 0.9676**: The accuracy has increased to 96.76%, indicating continued improvement in the model's performance.

### **Epoch 4/5**

**Output:**
```
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0876 - accuracy: 0.9734
```

**Explanation:**
- The model is on the fourth epoch.
- **2s**: The training time remains consistent.
- **loss: 0.0876**: The loss has decreased to 0.0876.
- **accuracy: 0.9734**: The accuracy has increased to 97.34%, indicating the model is performing even better.

### **Epoch 5/5**

**Output:**
```
1875/1875 [==============================] - 3s 1ms/step - loss: 0.0771 - accuracy: 0.9757
```

**Explanation:**
- The model is on the fifth and final epoch.
- **3s**: The time taken for this epoch is 3 seconds.
- **loss: 0.0771**: The loss has decreased to 0.0771.
- **accuracy: 0.9757**: The accuracy has increased to 97.57%, indicating strong performance on the training data.

### **Evaluation on Test Data**

**Output:**
```
313/313 [==============================] - 0s 742us/step - loss: 0.0745 - accuracy: 0.9780
[0.07453450560569763, 0.9779999852180481]
```

**Explanation:**
- **313/313**: The test dataset is divided into 313 batches.
- **0s**: The time taken to evaluate the test data is very quick, less than a second.
- **742us/step**: The average time per step (batch) is 742 microseconds.
- **loss: 0.0745**: The loss on the test data is 0.0745. This value is similar to the training loss, indicating that the model has not overfitted and generalizes well to unseen data.
- **accuracy: 0.9780**: The accuracy on the test data is 97.80%, showing that the model performs well on new, unseen data.
- **[0.07453450560569763, 0.9779999852180481]**: This final array contains the loss and accuracy on the test data. The first value is the loss, and the second value is the accuracy.

### **Summary**

The training and evaluation process shows that the model starts with a moderate accuracy and gradually improves over each epoch. By the end of the training, the model achieves high accuracy on both the training and test datasets, indicating effective learning and good generalization to new data. This process demonstrates how neural networks can be trained to perform classification tasks with high accuracy through iterative learning and evaluation.

### [Return to Main Page](../README.md)

## [Pytorch Fundamental Concepts](pytorch_fundamental_concepts.md)